client/core/trade.go

```diff
@@ -696,9 +843,31 @@ func (t *trackedTrade) swapMatches(matches []*matchTracker) error {
 
 	swaps.FeeRate = highestFeeRate
 
+	// Figure out how many have already been swapped.
+	var alreadySwapped int
+	for _, match := range t.matches {
+		if (match.Match.Side == order.Maker && match.Match.Status >= order.MakerSwapCast) ||
+			(match.Match.Side == order.Taker && match.Match.Status >= order.TakerSwapCast) {
+
+			alreadySwapped += 1
+		}
+	}
+	// If these are the last swaps, and the order is filled or canceled or
+	// revoked, then we don't need to lock the change coin.
+	isLastSwaps := len(matches)+alreadySwapped == len(t.matches)
+	isFilled := trade.Filled()+swappedQty == trade.Quantity
+	isCanceled := t.metaData.Status == order.OrderStatusCanceled || t.metaData.Status == order.OrderStatusRevoked
+	skipChange := isLastSwaps && (isFilled || isCanceled)
```

Comment:
> market or immediate limits won't have more matches added for this trade in the future, I believe.

Not quite. A market or immediate limit will be the taker, and will therefore be waiting on the maker to act and the transaction to have confirmations before swapping. If there are multiple makers, they may not all act at the same time, and there is no guarantee that all of their swaps are included in the same block. So even an immediate order might go through multiple `swapMatches`.

Reply:
I'm trying to understand the conditions that will prevent "final" market or immediate limits from having their change locked even though it's the last swap for them.  `isLastSwaps := len(matches)+alreadySwapped == len(t.matches)` is to know if `(*trackedTrade).swapMatches` will be called again for the currently known matches for this trade _and it's known matches_, and `isLastSwaps` tells of if there are more possibly more `swapMatches` to come for this trade (because the matches were made in `negotiate` before the takers starts waiting for the maker, as you've pointed out).  With market and immediate limit, I'm saying that `negotiate` (where `t.matches` inserts happen) will only ever happen once for such orders.

Reply:
Ah. I think I see. So if a non-standing order doesn't fill (or has unfillable remainder for market buy order), `isLastSwaps` will be `true`, but neither `isFilled` or `isCanceled` will ever be `true`. Will fix and test.

Reply:
Yeah, anything not completely filled and not on the book awaiting more matches... anything not booked anymore really.  `isCanceled || isFilled` covers standing limits.  But can we just do `skipChange := isLastSwaps && t.MetaData.Status != order.OrderStatusBooked`?  If the order is not booked, then that means there will be no more additions to `t.matches` via `negotiate`, right?

Reply:
Ah, you already did that, cool